### PR TITLE
fix: fix page navigation freeze with target-text reflow

### LINKED ISSUE #1856: Issue: Page navigation freezes with target-text reflow
Test case: [target-text-reflow-navigation/publication.json](https://raw.githubusercontent.com/vivliostyle/vivliostyle.js/master/packages/core/test/files/target-text-reflow-navigation/publication.json)

Test with canary viewer:
- https://vivliostyle.vercel.app/#src=https://raw.githubusercontent.com/vivliostyle/vivliostyle.js/master/packages/core/test/files/target-text-reflow-navigation/publication.json

Test with v2.40.0 viewer:
- https://vivliostyle.github.io/viewer/v2.40.0/#src=https://raw.githubusercontent.com/vivliostyle/vivliostyle.js/master/packages/core/test/files/target-text-reflow-navigation/publication.json

Both canary and v2.40.0 viewers have problems with page navigation.

- In canary viewer, when navigating from page 2 to the next page, the viewer freezes and becomes unresponsive.
  - In addition, the page number displayed in the page footer of page 2 is incorrectly shown as "3".
- In v2.40.0 viewer, when navigating from page 1 to the next page, error message "TypeError: Cannot read properties of undefined (reading 'spineIndex')" is displayed.

**Expected behavior:**

- The viewer does not freeze or show errors when navigating between pages.
- The page numbers are displayed correctly in the page footers, "1", "2", "3", "4" for the respective pages.
- The page numbers displayed in the TOC items should match the page numbers of their respective target pages, "3" or "4".

### PR DESCRIPTION
**Pull request by MurakamiShinyu**
When cross-document target-text() resolution causes a preceding spine (e.g. TOC) to expand by one or more pages, several issues occurred:

- Navigation from the expanded spine to the next spine froze the viewer because the pending layout position was never materialized into a page.
- Page counter values (counter(page)) in page margins of later spines were stale, showing incorrect numbers.
- target-counter() values in the TOC did not reflect the corrected page numbers after spine expansion.
- The viewer toolbar page slider showed wrong page numbers on direct navigation via TOC click or epubcfi.

Fix by:
- Saving counter state before popPageCounters in the cross-spine resolve path and restoring it for pending page rendering.
- Falling through to renderPage when a pending layout position exists but no page has been materialized.
- Deriving pageCounterOffset from the previous spine's last pageCounterStarts rather than the potentially stale global counter.
- After spine expansion, adjusting pageCountersById snapshots for later spines, shifting pageCounterStarts, patching counter(page) and target-counter() DOM nodes in already-rendered pages, and updating item.epage so the viewer slider reflects the correct position.

Closes #1856

Assisted-by: GitHub Copilot Chat

<details>
<summary>How the AI was used</summary>

- The human author described a page-navigation freeze and incorrect-page-number bug from `target-text()` reflow in detail, clarifying it was not a regression (it had never worked correctly, even in the older stable release) and pointing to a series of prior `target-text`/counter PRs as likely relevant context.
- Layout-regression testing surfaced a regression in existing target-counter/text tests from the first fix attempt; the human author worked with the AI to correct the counter values, catching a stale page-footer number after a TOC click, and later catching that TOC/epubcfi navigation had regressed further ("this is worse than before") — requiring another round of fixes.
- The human author accepted a cosmetic remaining limitation (toolbar page-slider off-by-one on direct navigation) flagged by a reviewer, directed the commit message, created the PR, addressed review comments, and reported and had fixed one more footer-page-number regression discovered after the fix was live.

</details>

### CODE REVIEW

File: packages/core/src/vivliostyle/epub.ts [hunk 1]
```diff
@@ -3089,11 +3192,30 @@ export class OPFView implements Vgen.CustomRendererFactory {
...
+          // Derive the page counter offset from the previous spine's last
+          // rendered page counter start rather than the global
+          // currentPageCounters, which may reflect stale state from a later
+          // spine that has been destroyed and is being recreated (e.g. after
+          // target-text reflow expanded an earlier spine).
+          const prevLastPageCounterStarts =
+            previousViewItem &&
+            previousViewItem.pageCounterStarts[
+              previousViewItem.pages.length - 1
+            ];
```
[the remark this answers is not in the document]
Comment (MurakamiShinyu) [reply packages/core/src/vivliostyle/epub.ts:3211]: I tried switching to pageCountersById (the post-render snapshot from finishPage()) but it caused a regression: during resolveUnresolvedReferencesForPage, cascade re-renders call finishPage() again, which overwrites pageCountersById entries with counter values from the resolve scope rather than the normal rendering scope. This made pageCounterOffset incorrect for sequential navigation (e.g., page 3 showed "5" in the footer).

pageCounterStarts is more stable here because preparePageCountersForRender preserves the stored value when oldPage && storedCounters is true, so it is not overwritten during resolve-scope re-renders. The + 1 adjustment accounts for the standard counter-increment: page 1 that updatePageCounters() applies.

You're right that an explicit counter-reset: page on the previous spine's last page could make the + 1 assumption inaccurate. However, this is an edge case that requires a separate fix (tracking the actual increment value from updatePageCounters), and I'd prefer to address it in a follow-up if a concrete test case surfaces. Updated the comment to explain why pageCounterStarts is used instead of pageCountersById.

File: packages/core/src/vivliostyle/counters.ts [hunk 2]
```diff
@@ -1158,6 +1158,87 @@ export class CounterStore {
...
+   * update their text content from the adjusted pageCountersById snapshots.
+   * Called after adjustPageCountersOfLaterSpines has shifted the counters.
+   */
+  updatePageCounterNodesInPages(pages: Vtree.Page[]): void {
+    for (const page of pages) {
+      if (!page?.container) continue;
+      // Get the adjusted counter values for this page from any tracked element
+      const anyId = Object.keys(page.elementsById)[0];
+      if (!anyId) continue;
+      const counters = this.pageCountersById[anyId];
```
[the remark this answers is not in the document]
Comment (MurakamiShinyu) [reply packages/core/src/vivliostyle/counters.ts]: Adopted the suggested approach: added a getPageCountersForPage() helper that first tries pageCountersById via tracked element IDs, then falls back to explicit per-page counters passed by the caller. updatePageCounterNodesInPages() now accepts an optional pageCountersByPage array parameter. The call site in epub.ts passes laterItem.pageCounterStarts as the fallback so pages without tracked element IDs can still be updated.
